hub/core/transform/transform.py

```diff
@@ -0,0 +1,166 @@
+            After transform is complete, ds_out has x samples in each tensor, where x is the number of outputs generated on applying pipeline on data_in.
+            2. All tensors are populated, but all have sampe length, say n. In this case all samples are simply appended to the dataset.
+            After transform is complete, ds_out has n + x samples in each tensor, where x is the number of outputs generated on applying pipeline on data_in.
+            3. All tensors are populated, but don't have the same length. This case is NOT supported.
+        pipeline_kwargs (List[dict], optional): A list containing extra arguments to be passed to the pipeline functions.
+            If this is None no extra arguments are passed to the pipeline functions.
+            If this contains exactly the same number of kwargs as number of functions in pipeline, corresponding kwargs are applied to each function.
+            If this contains less kwargs than the number of functions, only the starting functions upto the length of pipeline_kwargs get kwargs.
+            If this contains more kwargs than the number of functions, the extra kwargs are ignore.
+            Tip: If you want to apply kwargs to non-continuos functions, then use empty dictionaries in the middle.
+            For example:- transform(data_in, [fn1, fn2, fn3], ds_out, [{"a":5, "b":6}, {}, {"c":11, "s":7}]), only applies kwargs to fn1 and fn3.
+        scheduler (str): The scheduler to be used to compute the transformation.
+            Currently can be one of:-
+            threaded: Uses multithreading to perform the transform. Best applicable for I/O intensive transforms.
+            processed: Uses multiprocessing to perform the transform. Best applicable for CPU intensive transforms. Currently doesn't work with S3 or Hub cloud datasets.
+        workers (int): The number of workers to use for performing the transform. Defaults to 1.
+
+    Raises:
+        InvalidInputDataException: If ds_in passed to transform is invalid. It should support __getitem__ and __len__ operations.
+        TensorMismatchException: If one or more of the outputs generated during transform contain different tensors than the ones present in the output 'ds_out' provided to transform.
+        UnsupportedSchedulerException: If the scheduler passed is not recognized.
+        InvalidTransformException: If the output of any step in a transformation isn't dictionary or a list/tuple of dictionaries.
+    """
+    ds_out.flush()
+
+    if not hasattr(data_in, "__getitem__"):
+        raise InvalidInputDataException("__getitem__")
+    if not hasattr(data_in, "__len__"):
+        raise InvalidInputDataException("__len__")
+
+    # TODO: this check doesn't work currently. Will work once AL-1092 is merged and can be uncommented.
+    # for tensor in tensors:
+    #     assert len(ds_out[tensor]) == len(ds_out)
```

Comment:
should not have assertions outside of tests, please use exceptions

Reply:
but this shouldn't be inside of this method, should be done inside a test file

Reply:
No, this is not a test, this will check if the dataset provided to write transform is valid for transformation or not.
Will change to exception once length fix is pushed.